### PR TITLE
fix: Properly handling stalls due to 1-height differences

### PR DESCRIPTION
**Pull request by SkidanovAlex**
This change builds on top of Michael's changes to introduce `local_network` mode to stress.py

The nodes only learn about their peer's highest height via Block messages, and thus if a block message is lost and two nodes are exactly one height away from each other, the one behind will never learn it needs to catch up.
If now such a 1-height difference is split across two rouhgly even halves of the block producers, the system will stall.
    
Fixing it by re-broadcasting `head` if during some reasonable time the network made no progress.
    
With this change `local_network` mode passes in nayduck.
    
Test plan:
---------
Before this change stress.py local_network fails consistently due to chain stalling.
After:
http://nayduck.eastus.cloudapp.azure.com:3000/#/run/104
    
Out of five failures the first one is new (and needs debugging), but the chain is not stalled.
The remaining four are #3169

### CODE REVIEW

File: pytest/lib/proxy_instances.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    def __init__(self, reject_list, ordinal):
+        super().__init__(ordinal)
+        self.reject_list = reject_list
+
+    async def handle(self, msg, fr, to):
+        if fr in self.reject_list or to in self.reject_list:
+            msg_type = msg.enum if msg.enum != 'Routed' else msg.Routed.body.enum
+            logging.info(
+                f'NODE {self.ordinal} blocking message {msg_type} from {fr} to {to}')
```
Comment (SkidanovAlex) [inline pytest/lib/proxy_instances.py:15]: Compared to Michael's original PR I changed logging here to include the message type.

File: pytest/tests/stress/stress.py [hunk 2]
```diff
@@ -170,16 +170,29 @@ def monkey_node_restart(stopped, error, nodes, nonces):
...
+    last_time_height_updated = time.time()
+
     while stopped.value == 0:
+        _, cur_height = get_recent_hash(nodes[-1], 30)
+        if cur_height == last_height and time.time() - last_time_height_updated > 10:
+            time.sleep(25)
+        else:
+            last_height = cur_height
+            last_time_height_updated = time.time()
+            time.sleep(5)
```
Comment (SkidanovAlex) [inline pytest/tests/stress/stress.py:183]: This is also different from the original PR, in which it was a constant 9s sleep.
If the cluster got split into two halves with 1 height difference, it takes 8-9 seconds to even detect it, and then some time to recover, so sleeping longer when the cluster stalls.

File: chain/client/src/sync.rs [hunk 3]
```diff
@@ -52,19 +52,12 @@ pub const MAX_PENDING_PART: u64 = MAX_STATE_PART_REQUEST * 10000;
 pub const NS_PER_SECOND: u128 = 1_000_000_000;
 
 /// Get random peer from the hightest height peers.
-pub fn highest_height_peer(
-    highest_height_peers: &Vec<FullPeerInfo>,
-    min_height: BlockHeight,
```
Comment (bowenwang1996) [inline chain/client/src/sync.rs:57]: why is this removed?

File: pytest/tests/stress/stress.py [hunk 4]
```diff
@@ -558,16 +571,34 @@ def doit(s, n, N, k, monkeys, timeout):
...
+    monkey_names = [x.__name__ for x in monkeys]
+    proxy = None
+    logging.info(monkey_names)
+    if 'monkey_local_network' in monkey_names or 'monkey_global_network' in monkey_names:
+        assert config['local'], 'Network stress operations only work on local nodes'
+        reject_list = RejectListProxy.create_reject_list(1)
+        proxy = RejectListProxy(reject_list)
+        expect_network_issues()
+        block_timeout += 40
+        balances_timeout += 20
```
Comment (SkidanovAlex) [inline pytest/tests/stress/stress.py:583]: Also added 20s extra to wait for transactions to apply when the network is being messed up with.

**Comment by SkidanovAlex** [comment]
They don't need to sync if they are one block behind. Once the head is broadcasted from the nodes that are 1 height ahead, it will be accepted by the nodes that are 1 block behind, and thus syncing is not needed.

File: chain/client/src/sync.rs (hunk 3 above)
Comment (SkidanovAlex) [reply chain/client/src/sync.rs:57]: It was a workaround to make sure the syncing doesn't end prematurely (in the old approach once we start syncing, the code could choose a peer that is, same as us, one height behind and immediately stop syncing).
